perception/crosswalk_traffic_light_estimator/src/node.cpp

```diff
@@ -219,24 +219,157 @@ void CrosswalkTrafficLightEstimatorNode::updateLastDetectedSignal(
+  const TrafficSignalArray & msg)
 {
   const auto tl_reg_elems = crosswalk.regulatoryElementsAs<const lanelet::TrafficLight>();
 
+  std::unordered_map<lanelet::Id, size_t> valid_id2idx_map;  // detected traffic light
+
+  for (size_t i = 0; i < msg.signals.size(); ++i) {
+    auto signal = msg.signals[i];
+    valid_id2idx_map[signal.traffic_signal_id] = i;
+  }
+
   for (const auto & tl_reg_elem : tl_reg_elems) {
-    TrafficSignal output_traffic_signal;
-    TrafficSignalElement output_traffic_signal_element;
-    output_traffic_signal_element.color = color;
-    output_traffic_signal_element.shape = TrafficSignalElement::CIRCLE;
-    output_traffic_signal_element.confidence = 1.0;
-    output_traffic_signal.elements.push_back(output_traffic_signal_element);
-    output_traffic_signal.traffic_signal_id = tl_reg_elem->id();
-    msg.signals.push_back(output_traffic_signal);
+    auto id = tl_reg_elem->id();
+    // if valid prediction exists, overwrite the estimation; else, use the estimation
+    if (valid_id2idx_map.count(id)) {
+      size_t idx = valid_id2idx_map[id];
+      auto signal = msg.signals[idx];
+      isFlashing(signal);  // check if it is flashing
+      // update output msg according to flashing and current state
+      output.signals[idx].elements[0].color = updateState(signal);
+    } else {
+      TrafficSignal output_traffic_signal;
+      TrafficSignalElement output_traffic_signal_element;
+      output_traffic_signal_element.color = color;
+      output_traffic_signal_element.shape = TrafficSignalElement::CIRCLE;
+      output_traffic_signal_element.confidence = 1.0;
+      output_traffic_signal.elements.push_back(output_traffic_signal_element);
+      output_traffic_signal.traffic_signal_id = id;
+      output.signals.push_back(output_traffic_signal);
+    }
+  }
+}
+
+void CrosswalkTrafficLightEstimatorNode::isFlashing(const TrafficSignal & signal)
+{
+  const auto id = signal.traffic_signal_id;
+
+  // no record of detected color in last_detect_color_hold_time_(2.0s)
+  if (is_flashing_.count(id) == 0) {
+    is_flashing_.insert(std::make_pair(id, false));
+    return;
+  }
+
+  // flashing green
+  if (
+    signal.elements.front().color == TrafficSignalElement::UNKNOWN &&
+    signal.elements.front().confidence != 0 &&  // not due to occlusion
+    current_state_.at(id) != TrafficSignalElement::UNKNOWN) {
+    is_flashing_.at(id) = true;
+    return;
+  }
+
+  // history exists
+  if (last_colors_.count(id) > 0) {
+    std::vector<TrafficSignalAndTime> history = last_colors_.at(id);
+    for (const auto & h : history) {
+      if (h.first.elements.front().color != signal.elements.front().color) {
+        // keep the current value if not same with input signal
+        return;
+      }
+    }
+    // all history is same with input signal
+    is_flashing_.at(id) = false;
+  }
+
+  // no record of detected color in last_color_hold_time_(1.0s)
```

Comment:
```suggestion
  // no record of detected color in last_color_hold_time_
```
https://github.com/autowarefoundation/autoware.universe/pull/5871/files#r1433533089